extensions/oidc/runtime/src/main/java/io/quarkus/oidc/runtime/CodeAuthenticationMechanism.java

```diff
@@ -268,102 +284,82 @@ private boolean shouldAutoRedirect(TenantConfigContext configContext, RoutingCon
-
-                                            if (!removeRedirectParams) {
-                                                finalUriWithoutQuery.append('?').append(absoluteUri.getRawQuery());
-                                            }
-                                            if (finalUserQuery != null) {
-                                                finalUriWithoutQuery.append(!removeRedirectParams ? "" : "?");
-                                                finalUriWithoutQuery.append(finalUserQuery);
-                                            }
-                                            String finalRedirectUri = finalUriWithoutQuery.toString();
-                                            LOG.debugf("Final redirect URI: %s", finalRedirectUri);
-                                            uniEmitter.fail(new AuthenticationRedirectException(finalRedirectUri));
-                                        } else {
-                                            uniEmitter.complete(augmentIdentity(identity, opaqueAccessToken,
-                                                    opaqueRefreshToken, context));
-                                        }
-                                    }
-                                }, new Consumer<Throwable>() {
-                                    @Override
-                                    public void accept(Throwable throwable) {
-                                        uniEmitter.fail(throwable);
-                                    }
-                                });
+
+        Uni<AuthorizationCodeTokens> codeFlowTokensUni = getCodeFlowTokensUni(context, configContext, code);
+
+        return codeFlowTokensUni
+                .onItemOrFailure()
+                .transformToUni(new BiFunction<AuthorizationCodeTokens, Throwable, Uni<? extends SecurityIdentity>>() {
+                    @Override
+                    public Uni<SecurityIdentity> apply(final AuthorizationCodeTokens tokens, final Throwable t) {
+                        return Uni.createFrom().emitter(new Consumer<UniEmitter<? super SecurityIdentity>>() {
+                            @Override
+                            public void accept(UniEmitter<? super SecurityIdentity> emitter) {
+                                if (t != null) {
+                                    LOG.debugf("Exception during the code to token exchange: %s", t.getMessage());
+                                    emitter.fail(new AuthenticationCompletionException(t));
+                                } else {
+                                    context.put(NEW_AUTHENTICATION, Boolean.TRUE);
+                                    context.put(CODE_FLOW_ACCESS_TOKEN, tokens.getAccessToken());
+                                    authenticate(identityProviderManager, new IdTokenCredential(tokens.getIdToken(), context))
+                                            .subscribe().with(new Consumer<SecurityIdentity>() {
```

Comment:
I believe we can simplify this a bit, but that's not the focus of this PR.

Reply:
@cescoffier That is a good point - I've refactored it the same way the re-authentication is done a few blocks above, with `.map` and `transform` for the exception, and it looks much better, going to push a bit later